[WIP] Add feature to restrict user access for guests.

sahil839 opened the pull request:
<!-- Describe your pull request here.-->
- Prep commit to remove `select_related` call.
- Adds the new setting.
- Pass "realm__can_access_all_users_group" to `select_related` calls.
- Pass bogus data for inaccessible users.
- Update GET user endpoints tor return errors for inaccessible users.
- Update user creation events sent to guests.
- Do not send user update events to users who cannot access modified user.
- Update code for presence and status updates.
- Send user creation events on adding subscriptions.
- Send user remove events when unsubscribing.
- Restrict sending DMs to inaccessible users
- Send user creation events on sending DMs.
- Restrict accessing avatar using endpoint for inaccessible users.

Fixes: <!-- Issue link, or clear description.--> #10970

<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->

<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [ ] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [ ] Explains differences from previous plans (e.g., issue description).
- [ ] Highlights technical choices and bugs encountered.
- [ ] Calls out remaining decisions and concerns.
- [ ] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [ ] Each commit is a coherent idea.
- [ ] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [ ] Visual appearance of the changes.
- [ ] Responsiveness and internationalization.
- [ ] Strings and tooltips.
- [ ] End-to-end functionality of buttons, interactions and flows.
- [ ] Corner cases, error conditions, and easily imagined bugs.
</details>

sahil839 commented on zerver/models.py:
```diff
@@ -4077,9 +4081,9 @@ def get_user_by_delivery_email(email: str, realm: Realm) -> UserProfile:
-        delivery_email__iexact=email.strip(), realm=realm
-    )
+    return UserProfile.objects.select_related(
+        "realm", "realm__can_access_all_users_group", "bot_owner"
```
This change helps us in saving a query as we access the "can_access_all_users_group" field at multiple places where realm object is fetched from user object.

sahil839 commented on zerver/lib/avatar.py:
```diff
@@ -147,3 +147,7 @@ def is_avatar_new(ldap_avatar: bytes, user_profile: UserProfile) -> bool:
+
+
+def get_avatar_for_inaccessible_user() -> str:
+    return staticfiles_storage.url("images/default-avatar.png")
```
I have used this avatar just for now, as we had not decided the color before. Will update it soon.

timabbott replied:
OK

sahil839 commented on zerver/lib/users.py:
```diff
@@ -476,6 +486,125 @@ def format_user_row(
+        return True
+
+    realm = acting_user.realm
+    if is_user_in_group(realm.can_access_all_users_group, acting_user):
```
We can directly check whether the group is set to "Everyone" here because we only provide two options here currently.

sahil839 commented on zerver/lib/users.py:
```diff
@@ -476,6 +486,125 @@ def format_user_row(
+
+
+def get_subscribers_of_target_user_subscriptions(
+    target_users: List[UserProfile],
```
This (and the below function) are written to support querying common subscriptions but we use this same for the cases where we only need to query for single user to avoid having similar queries at two places. Querying multiple users using single query is mainly used to  query users for sending user creation and remove events.

sahil839 commented on zerver/lib/users.py:
```diff
@@ -575,6 +774,13 @@ def get_raw_user_data(
+            # delivery_email field is None only for inaccessible users.
+            # We already have the required data for them in row object,
+            # so we can just add it to result and continue.
+            result[row["user_id"]] = row
```
This check using `delivery_email` can be confusing because we actually pass it as `None` for users who cannot access email, but it is done later in the code in `format_user_row` and here it always includes real email.

timabbott replied:
Hmm, this feels really messy. I think a potentially good strategy would be a preparatory refactoring PR that:

* Renames `get_raw_user_data` to `get_users_for_api`
* Creates a `TypedDict` called, say,  `RawUserDict` for the return type of `user_profile_to_user_row` and `get_realm_user_dicts`; and another, `APIUserDict` for the type that `format_user_row` returns, and another for the type that `get_realm_user_dicts` returned.

And then the new `get_user_dicts_in_realm` could return a `Union[APIUserDict, RawUserDict]`, and this block is just storing in `result` a `APIUserDict` if that's what we got back from `get_user_dicts_in_realm`.

Or maybe it'd be better to change `get_user_dicts_in_realm` to return a 2-tuple, one that's a list of `RawUserDict`s and the other than's a list of `APIUserDict`s? But in any case, I think defining types here would allow us to not do this hacky `delivery_email` check.

timabbott replied:
This function now reads really nicely, thanks!

sahil839 commented on zerver/tests/test_signup.py:
```diff
@@ -935,7 +935,7 @@ def test_register(self) -> None:
         # to sending messages, such as getting the welcome bot, looking up
         # the alert words for a realm, etc.
-        with self.assert_database_query_count(103), self.assert_memcached_count(18):
+        with self.assert_database_query_count(104), self.assert_memcached_count(18):
```
This addition in query is for accessing `realm.can_access_all_users_group.name`. Since the realm object is obtained from `PreregistrationUser` obejct, there is no simple way of using `select_related` currently and may require some more refactoring.

timabbott replied:
It's probably fine to just have this one extra query for user creation. But please add a `TODO` comment about this detail just before the `user_access_restricted_in_realm` call in this commit noting that it ends up doing a database query because the queries to fetch `PreregistrationUser` do not prefetch it.

sahil839 commented on zerver/tests/test_users.py:
```diff
@@ -830,7 +830,7 @@ def test_create_user_with_multiple_streams(self) -> None:
         prereg_user = PreregistrationUser.objects.get(email="fred@zulip.com")
 
-        with self.assert_database_query_count(93):
+        with self.assert_database_query_count(94):
```
This increase in query is also because of the reason mentioned in https://github.com/zulip/zulip/pull/27544/commits/b2b3594ac11e3ff549a96ab76b7ceeabd759497c#r1382436152.

sahil839 commented on zerver/lib/users.py:
```diff
@@ -578,16 +601,21 @@ def get_subscribers_of_target_user_subscriptions(
+    subs_in_target_user_subscriptions_query = Subscription.objects.filter(
+        recipient_id__in=list(target_users_subbed_recipient_ids),
+        active=True,
+        recipient__type__in=[Recipient.STREAM, Recipient.HUDDLE],
```
Adding code to include only active users here makes this function somewhat complicated. One possible workaround would be to filter out inactive users in callers of these functions wherever required in a simpler way like just taking intersection with `active_user_ids(realm.id)` and in some cases this would not be required since we already filter out inactive users.

But having an option to exclude deactivated users seems better logically to me even though it makes it somewhat hard to read.

sahil839 commented on zerver/lib/users.py:
```diff
@@ -860,7 +908,14 @@ def is_2fa_verified(user: UserProfile) -> bool:
+        active_users = user_profile.realm.get_active_users()
+    else:
+        user_ids_who_can_access_user = get_user_ids_who_can_access_user(user_profile)
+        active_users = UserProfile.objects.filter(
```
We could have just directly got user objects here instead of first getting IDs and then objects, but we use the same code which is used for getting accessible users for multiple users, so that makes it somewhat inefficient to query large number of `UserProfile` objects.

And this extra query should not be much problematic as we need user objects only when changing emails and email visibility setting.

timabbott replied:
This makes sense to me.

sahil839 commented on zerver/tests/test_invite.py:
```diff
@@ -108,7 +108,7 @@ def test_query_counts_for_new_user_using_default_streams(self) -> None:
         new_user = self.create_simple_new_user(realm, "alice@zulip.com")
 
-        with self.assert_database_query_count(11):
+        with self.assert_database_query_count(12):
```
Here also query count increases just because of accessing `realm.can_access_all_users_group.name`.

sahil839 commented on zerver/actions/message_send.py:
```diff
@@ -1557,6 +1618,10 @@ def check_message(
         check_private_message_policy(realm, sender, user_profiles)
 
+        recipients_for_user_creation_events = get_recipients_for_user_creation_events(
+            realm, sender, user_profiles
```
We compute this early here because huddle objects are created before sending a message and thus we would not get the correct output of accessible users after creation the huddle objects. But this means we have to pass this data to a couple of functions below in the stack.

timabbott replied:
OK; let's put a version of this note into a comment just before the call. (In general, I think when posting a GitHub comment like this, it's always worth considering whether it makes sense to copy it into the code.

sahil839 commented on zerver/tests/test_events.py:
```diff
@@ -536,6 +537,31 @@ def test_huddle_send_message_events(self) -> None:
+        )
+        check_direct_message("events[0]", events[0])
+        check_realm_user_add("events[1]", events[1])
+        check_realm_user_add("events[2]", events[2])
```
The order of events here in tests is not same as in we would get in production environment because `notify_created_user` uses `send_event_on_commit` and the tests code execute the `on_commit` calls at last.

timabbott replied:
I think actually the same thing will happen in production -- the event sent by `send_event_on_commit` will happen last. We likely need to change the main direct message event to be sent via `send_event_on_commit` rather than `send_event` to fix this bug.

sahil839 commented:
There is still a lot of changes to be done in this PR like API docs changes, adding client compatibility field, avatar for unknown users, but that should be quick and not take much time and in the meantime can have feedback mainly on the DB queries used and the approach.

timabbott commented on zerver/lib/users.py:
```diff
@@ -476,6 +486,125 @@ def format_user_row(
+
+def check_user_can_access_all_users(acting_user: Optional[UserProfile]) -> bool:
+    if acting_user is None:
+        return True
```
I guess this is the spectators case; can you add a comment on why this value is `True`?

sahil839 replied:
Yes, we [discussed](https://chat.zulip.org/#narrow/stream/101-design/topic/access.20to.20all.20users.20in.20realm.20.2310970/near/1644685) about allowing access to all the users for spectators. Would add a comment here.

timabbott commented on zerver/lib/users.py:
```diff
@@ -511,6 +515,45 @@ def check_user_can_access_all_users(acting_user: Optional[UserProfile]) -> bool:
+) -> bool:
+    if check_user_can_access_all_users(user_profile) or not user_access_restricted_in_realm(
+        target_user
+    ):
```
We should do these checks in the opposite order, in that it'll save a `is_user_in_group` query to do do so.

Also probably would be better to not put them on the same line of code.

sahil839 replied:
Updated this.

timabbott commented on zerver/lib/users.py:
```diff
@@ -511,6 +515,45 @@ def check_user_can_access_all_users(acting_user: Optional[UserProfile]) -> bool:
+    ):
+        return True
+
+    assert user_profile is not None
```
Presumably this just reflects that we haven't tried to make this work for spectators yet, right?

sahil839 replied:
We allow spectators access to all users so `if check_user_can_access_all_users` would return true.

timabbott replied:
Yeah. I can imagine a world post client capability where we start trying to use this code path with spectators for performance reasons, but we can certainly leave that for later.

timabbott commented on zerver/lib/users.py:
```diff
@@ -511,6 +515,45 @@ def check_user_can_access_all_users(acting_user: Optional[UserProfile]) -> bool:
+        Q(sender_id=target_user.id, recipient_id=user_profile.recipient_id)
+        | Q(recipient_id=target_user.recipient_id, sender_id=user_profile.id)
+    ).exists():
+        return True
```
Let's add a TODO comment somewhere around here that going to the `Message` table is expensive, so we do it last.

(Long-term, I think we will probably move 1:1 DMs to use the `Huddle` table, and thus be able to do this whole process using only the `Subscription`/`Recipient` queries above).

timabbott commented on zerver/lib/users.py:
```diff
@@ -511,6 +515,45 @@ def check_user_can_access_all_users(acting_user: Optional[UserProfile]) -> bool:
+    assert user_profile is not None
+
+    if target_user.id == user_profile.id:
+        return True
```
We probably will need a bonus special case for system bots here, so that one can mute them, right?

(I can't remember if we made that work elsewhere, but this seems logical).

sahil839 replied:
The system bots will also have `is_bot` field set to `True` so `user_access_restricted_in_realm` should hanlde that I guess.

timabbott commented on zerver/tests/test_users.py:
```diff
@@ -2545,6 +2556,47 @@ def test_restricted_access_to_users(self) -> None:
+        # because he is not subscribed to the stream now.
+        # 2. cordelia, because the guest user is not subscribed
+        # to Verona anymore to which cordelia is subscribed.
+        inaccessible_users = [othello, cordelia, desdemona]
```
This comment seems to incorrectly not cover Desdemona?

sahil839 replied:
Fixed it.

timabbott commented on zerver/tests/test_users.py:
```diff
@@ -2545,6 +2556,47 @@ def test_restricted_access_to_users(self) -> None:
+        members_group = UserGroup.objects.get(name="role:members", realm=realm)
+        do_change_realm_permission_group_setting(
+            realm, "can_access_all_users_group", members_group, acting_user=None
+        )
```
I think these two lines are already present in `set_up_db_for_testing_user_access`

sahil839 replied:
Removed those lines.

timabbott commented on zerver/lib/users.py:
```diff
@@ -524,6 +653,76 @@ def get_cross_realm_dicts() -> List[Dict[str, Any]]:
+        is_billing_admin=False,
+        role=UserProfile.ROLE_MEMBER,
+        is_bot=False,
+        full_name="Unknown user",
```
We should probably tag "Unknown user" for translation, right?

sahil839 replied:
Done this. Also added a constant which is set to "Unknown user".

timabbott commented on zerver/lib/users.py:
```diff
@@ -524,6 +653,76 @@ def get_cross_realm_dicts() -> List[Dict[str, Any]]:
+        {user_profile.id}
+        | subscribers_dict_of_target_user_subscriptions[user_profile.id]
+        | users_involved_in_dms_dict[user_profile.id]
+    )
```
We should probably add a comment noting that this doesn't include cross-realm bots and why that's correct (presumably, because they're sent elsewhere in the API).

sahil839 replied:
Added a comment here about bots, but we would need to include bots here as well by using `include_bots` parameter.

timabbott commented on zerver/tests/test_users.py:
```diff
@@ -2504,6 +2508,43 @@ def test_user_email_according_to_email_address_visibility_setting(self) -> None:
+            realm, "can_access_all_users_group", members_group, acting_user=None
+        )
+        self.login("polonius")
+        result = orjson.loads(self.client_get("/json/users").content)
```
Let's capture the number of queries for this operation and assert it -- it's likely to be fairly expensive.

sahil839 replied:
Updated the test.

timabbott commented on zerver/actions/create_user.py:
```diff
@@ -320,13 +322,25 @@ def notify_created_user(user_profile: UserProfile) -> None:
+            if user.is_guest:
+                user_ids_without_access_to_created_user.append(user.id)
+            else:
+                users_with_access_to_created_users.append(user)
```
This block should have a comment:
# This logic assumes that `can_access_all_users_group` is EVERYONE or MEMBERS.

sahil839 replied:
Added the comment.

timabbott commented on zerver/actions/create_user.py:
```diff
@@ -359,6 +373,14 @@ def notify_created_user(user_profile: UserProfile) -> None:
+        event = dict(
+            type="realm_user",
+            op="add",
+            person=get_data_for_inaccessible_user(user_profile.id, True, user_profile.realm),
```
This feels hard to read. We should change the parameter order and use mandatory kwargs, for example: `get_data_for_inaccessible_user(user_profile.realm, user_profile.id, user_is_active=True)`

sahil839 replied:
We have removed the `is_active` parameter as we decided to pass it as `True` for all users.

timabbott commented on zerver/lib/users.py:
```diff
@@ -476,6 +492,204 @@ def format_user_row(
+        | users_in_subscribed_streams_or_huddles_dict[target_user.id]
+        | users_involved_in_dms_dict[target_user.id]
+    )
+    return list(user_ids_who_can_access_target_user)
```
I think the structure for this PR is pretty good. But I expect that `get_users_involved_in_dms_with_target_users` with its queries on the `Message` table, is, I think, far too expensive for a function that will be called often from presence code paths for every user in an organization that has this feature enabled.

I'm not sure what to do about it. I suppose one theory would be that we could work towards creating `Huddle` objects with `Subscription`s for all 1:1 DM conversations (even if we don't do the more complex refactoring that would be required to create corresponding `Recipient` objects and then repoint `Message` objects at them without breaking anything). In other words, doing the first part of the migrations needed to do https://github.com/zulip/zulip/issues/25713.

With that implemented, it would become the case that we at least only need to query the `Subscription`/`Recipient` tables in order to do these checks. That might still be too slow ... but it at least would be a much simpler problem statement to optimize.

sahil839 commented on zproject/default_settings.py:
```diff
@@ -587,3 +587,7 @@
+
+# We recommend to set this as False for organizations with more
+# than 100 users.
+LIMIT_PRESENCE_VISIBILITY_FOR_GUESTS = False
```
I am not sure what should be the default value for this. Also, not sure regarding the name, do we want to use `LIMIT_PRESENCE_VISIBILITY_FOR_LIMITED_GUESTS` (we came up with "limited guest" when we were thinking of adding a separate role, but we are not doing it now). Or we can have `LIMIT_PRESENCE_VISIBILITY_FOR_RESTRICTED_GUESTS`.

timabbott replied:
I think `CAN_ACCESS_ALL_USERS_GROUP_LIMITS_PRESENCE` would be an OK name.

sahil839 commented on zerver/views/presence.py:
```diff
@@ -48,6 +49,11 @@ def get_presence_backend(
     if target.is_bot:
         raise JsonableError(_("Presence is not supported for bot users."))
 
+    if settings.LIMIT_PRESENCE_VISIBILITY_FOR_GUESTS and not check_can_access_user(
```
I don't think we need to take care of loading time here as this endpoint is not used by webapp as of now, but just added the condition to check server level setting for consistency.

timabbott replied:
Agreed on both points that this makes sense.

sahil839 commented:
Updated the PR as per feedback. Posted the replies to comments above. Also added some changes regarding how to handle deactivated users as discussed on CZO.

Some changes that I working on now, excluding the client part -
- Updating the avatar file. This is being discussed.
- Sending user remove events on user being deactivated as we do not consider deactivated users as stream subscribers. We do not send `peer_remove` events on deactivation, otherwise it would have been much simpler to hande it.
- Updating the messages (in message events and message fetch endpoints) and push notifications data to not contain real data of inaccessible users.
- Adding a client capability.

timabbott commented on zerver/lib/users.py:
```diff
@@ -548,6 +671,74 @@ def get_cross_realm_dicts() -> List[APIUserDict]:
 
 
+def get_data_for_inaccessible_user(realm: Realm, user_id: int) -> APIUserDict:
+    user_email = Address(username=f"user{user_id}", domain=get_fake_email_domain(realm)).addr_spec
```
Renamed this variable to `fake_email` to make it more obvious we're not sending a real email here.

timabbott commented on zerver/lib/users.py:
```diff
@@ -501,6 +510,120 @@ def format_user_row(
+
+    subs_in_target_user_subscriptions = subs_in_target_user_subscriptions_query.order_by(
+        "recipient_id"
+    ).values("user_profile_id", "recipient_id")
```
We will probably want a specialized version of this query for the single-target-user case, since that's what we use in the high throughput code paths like presence. The key optimization is that that we can do something like `.values_list("user_profile_id", flat=True).distinct()` -- i.e. not sending multiple copies from the database to Python for a single `recipient_id`

timabbott replied:
But I think this can be a follow-up optimization; going to merge through this commit so we can get the migrations integrated.

timabbott replied:
@sahil839 is this optimization implemented? Wondering if we need a follow-up issue.

timabbott commented:
I think the first 5 commits look good. I pulled them out to a preparatory PR: https://github.com/zulip/zulip/pull/27660

sahil839 commented on zerver/lib/users.py:
```diff
@@ -526,6 +542,51 @@ def check_user_can_access_all_users(acting_user: Optional[UserProfile]) -> bool:
+        recipient_id__in=subscribed_recipient_ids,
+        user_profile=target_user,
+        active=True,
+        is_user_active=True,
```
One thing as mentioned in the commit message is that we would not allow guests to mute or unmute deactivated users who were part of the streams that the guest is subscribed to but were not involved in DMs with them.

timabbott replied:
The muting aspect is weird but likely not actually important. I'll add a comment to keep track of this observation.

sahil839 commented on zerver/openapi/zulip.yaml:
```diff
@@ -1133,6 +1133,13 @@ paths:
+                                who can access the user who came back online.
+
+                                **Changes**: Prior to Zulip 8.0 (feature level 227), this event
+                                was sent to all users in the organization.
```
I am not much sure about this changes, and might be phrased in a better way.

timabbott replied:
This is probably OK but might deserve whatever editing pass we apply to the other discussion of accessing users.

sahil839 commented on zerver/openapi/zulip.yaml:
```diff
@@ -1036,8 +1036,13 @@ paths:
-                                user joins. Processing this event is important to being able
-                                to display basic details on other users given only their ID.
+                                user joins or when a guest user gains access to a user.
+                                Processing this event is important to being able to display
```
Should we also describe the cases when a guest will gain access to a user?

timabbott replied:
I think so but possibly in /help/ documentation linked to from here; see https://github.com/zulip/zulip/pull/27544#discussion_r1401579684.

sahil839 commented on zerver/actions/streams.py:
```diff
@@ -893,6 +894,64 @@ def send_subscription_remove_events(
+                event_remove_user = dict(
+                    type="realm_user",
+                    op="remove",
+                    person=dict(user_id=user_id, full_name=str(UserProfile.INACCESSIBLE_USER_NAME)),
```
I think we can remove the `full_name` field here since I guess existing clients do not use this event (need to recheck though). Have kept it for now as we used to send it in the deprecated event.

timabbott replied:
Yeah, I expect we can remove that. Please post in "api design" about this to confirm that theory.

sahil839 commented on zerver/openapi/zulip.yaml:
```diff
@@ -1085,7 +1085,9 @@ paths:
-                                No longer sent by modern Zulip servers.
+                                Event sent to guest users when they lose access to a user.
+
+                                **Changes**: Prior to Zulip 8.0 (feature level 227), this event was deprecated.
```
Tbh, I am not sure how to correctly update the docs here.

timabbott replied:
This seems OK for this PR; though we should probably discuss in "api documentation" polishing these docs now that all these corner cases are implemented. The first step might be adding a /help/ article that we can link to about the concept of "losing access to a user" text, though; @alya FYI.

sahil839 commented on zerver/actions/users.py:
```diff
@@ -245,6 +252,65 @@ def change_user_is_active(user_profile: UserProfile, value: bool) -> None:
+        recipient__type__in=[Recipient.STREAM, Recipient.HUDDLE],
+        is_user_active=True,
+        active=True,
+    ).values_list("recipient__type", "user_profile_id")
```
Did not use existing functions here because we need separate subscriptions for streams and huddles.

timabbott replied:
Added this comment
```
+    # This code path is parallel to
+    # get_subscribers_of_target_user_subscriptions, but can't reuse it
+    # because we need to process stream and huddle subscriptions
+    # separately.
```

sahil839 commented:
Updated the PR.
- Added a couple of commits to send remove events when deactivating user and streams.
- Updated API docs as well, but might still need some more changes.
- Added a commit to fix `apply_event` code.
- Updated the avatar used for inaccessible users.

One test is failing in some of the workflows and not others. It also passes locally for me.

sahil839 commented:
Fixed the tests.

timabbott commented on zerver/actions/message_send.py:
```diff
@@ -1460,6 +1461,21 @@ def check_private_message_policy(
+    if check_user_can_access_all_users(sender):
+        return
+
+    users_accessible_to_sender = set(get_accessible_user_ids(realm, sender))
```
`get_accessible_user_ids` is quite expensive for this check; in the common case, `user_profiles` is going to be 1-2 users, and it should in theory be much cheaper to check facts like whether we've had any DMs with those users specifically than fetching the set of all users accessible to the sender.

Since sending DMs is a frequent operation, we probably do care about the latency this introduces for DMs sent by limited guests. It likely requires some messy plumbing, but I would expect a `user_profile__in=[small list]` filter on `subs_in_target_user_subscriptions_query` to be fairly significant too.

I think this is OK to fix in a follow-up; but I might add a TODO comment here before merging.

timabbott replied:
Opened https://github.com/zulip/zulip/issues/27835 to track this.

timabbott commented on zerver/lib/event_schema.py:
```diff
@@ -414,12 +414,11 @@ def check_heartbeat(
 )
 
-message_fields = [
+baisc_message_fields = [
```
`basic_message_fields` has a typo.

timabbott replied:
Fixed, pushing shortly.

timabbott commented on zerver/tests/test_users.py:
```diff
@@ -826,7 +826,7 @@ def test_create_user_with_multiple_streams(self) -> None:
         prereg_user = PreregistrationUser.objects.get(email="fred@zulip.com")
 
-        with self.assert_database_query_count(94):
+        with self.assert_database_query_count(95):
```
Why does this add an extra database query? This commit seems like it would not do any extra queries unless it's for `realm.can_access_all_users_group.name == SystemGroups.EVERYONE`?

timabbott replied:
Indeed, changing https://github.com/zulip/zulip/pull/27544#discussion_r1401561350 to move the "are there any guests" check up caused this query to no longer be required.

timabbott commented on zerver/actions/message_send.py:
```diff
@@ -1476,6 +1494,55 @@ def check_sender_can_access_recipients(
+
+    guest_recipients = [user for user in user_profiles if user.is_guest]
+    if len(guest_recipients) == 0:
+        return recipients_for_user_creation_events
```
Let's add a comment noting that this block relies on the assumption that `realm.can_access_all_users_group` contains all `MEMBERS`. Maybe:

> If none of the users in the direct message conversation are guests, then there is no possible `realm.can_access_all_users_group` policy that would mean sending this message changes any user's user access to other users.

We might also want to consider putting this check before the `realm.can_access_all_users_group` one, since that might save a database query when `realm.can_access_all_users_group` is not prefetched?

timabbott replied:
Applied.

timabbott commented on zerver/actions/streams.py:
```diff
@@ -74,6 +74,38 @@
+def send_user_remove_events_on_stream_deactivation(
+    stream: Stream, subscribed_users: List[UserProfile]
+) -> None:
+    guest_subscribed_users = [user for user in subscribed_users if user.is_guest]
```
Should this exit early if `guest_subscribed_users` was empty?

timabbott replied:
Added such a check. It's not important given that stream deactivations are rare and expensive, though.

timabbott commented:
I made a few changes, mostly adding comments but also resolving a few minor review comments noted above. With the below changes, I've marked this to merge once CI passes.

``` diff
diff --git a/zerver/actions/message_send.py b/zerver/actions/message_send.py
index 005a355b9b..b51b1ed871 100644
--- a/zerver/actions/message_send.py
+++ b/zerver/actions/message_send.py
@@ -1505,6 +1505,14 @@ def get_recipients_for_user_creation_events(
     """
     recipients_for_user_creation_events: Dict[UserProfile, Set[int]] = defaultdict(set)
 
+    # If none of the users in the direct message conversation are
+    # guests, then there is no possible can_access_all_users_group
+    # policy that would mean sending this message changes any user's
+    # user access to other users.
+    guest_recipients = [user for user in user_profiles if user.is_guest]
+    if len(guest_recipients) == 0:
+        return recipients_for_user_creation_events
+
     if realm.can_access_all_users_group.name == SystemGroups.EVERYONE:
         return recipients_for_user_creation_events
 
@@ -1513,10 +1521,6 @@ def get_recipients_for_user_creation_events(
             recipients_for_user_creation_events[sender].add(user_profiles[0].id)
         return recipients_for_user_creation_events
 
-    guest_recipients = [user for user in user_profiles if user.is_guest]
-    if len(guest_recipients) == 0:
-        return recipients_for_user_creation_events
-
     users_involved_in_dms = get_users_involved_in_dms_with_target_users(guest_recipients, realm)
     subscribers_of_guest_recipient_subscriptions = get_subscribers_of_target_user_subscriptions(
         guest_recipients
diff --git a/zerver/actions/streams.py b/zerver/actions/streams.py
index 85c39b4c83..cc2c5727d9 100644
--- a/zerver/actions/streams.py
+++ b/zerver/actions/streams.py
@@ -78,6 +78,12 @@ def send_user_remove_events_on_stream_deactivation(
     stream: Stream, subscribed_users: List[UserProfile]
 ) -> None:
     guest_subscribed_users = [user for user in subscribed_users if user.is_guest]
+
+    if len(guest_subscribed_users) == 0:
+        # Save a few database queries in the case that the stream
+        # didn't contain any guest users.
+        return
+
     other_subscriptions_of_subscribed_users = get_subscribers_of_target_user_subscriptions(
         guest_subscribed_users
     )
diff --git a/zerver/actions/users.py b/zerver/actions/users.py
index a7268d60d6..2300d59567 100644
--- a/zerver/actions/users.py
+++ b/zerver/actions/users.py
@@ -267,6 +267,10 @@ def send_events_for_user_deactivation(user_profile: UserProfile) -> None:
     non_guest_user_ids = active_non_guest_user_ids(realm.id)
     users_involved_in_dms_dict = get_users_involved_in_dms_with_target_users([user_profile], realm)
 
+    # This code path is parallel to
+    # get_subscribers_of_target_user_subscriptions, but can't reuse it
+    # because we need to process stream and huddle subscriptions
+    # separately.
     deactivated_user_subs = Subscription.objects.filter(
         user_profile=user_profile,
         recipient__type__in=[Recipient.STREAM, Recipient.HUDDLE],
diff --git a/zerver/lib/event_schema.py b/zerver/lib/event_schema.py
index 81194c5b2c..1fe7f3b172 100644
--- a/zerver/lib/event_schema.py
+++ b/zerver/lib/event_schema.py
@@ -414,7 +414,7 @@ _check_topic_links = DictType(
     ]
 )
 
-baisc_message_fields = [
+basic_message_fields = [
     ("avatar_url", OptionalType(str)),
     ("client", str),
     ("content", str),
@@ -435,7 +435,7 @@ baisc_message_fields = [
 ]
 
 message_fields = [
-    *baisc_message_fields,
+    *basic_message_fields,
     ("display_recipient", str),
     ("stream_id", int),
 ]
@@ -459,7 +459,7 @@ _check_direct_message_display_recipient = DictType(
 )
 
 direct_message_fields = [
-    *baisc_message_fields,
+    *basic_message_fields,
     ("display_recipient", ListType(_check_direct_message_display_recipient)),
 ]
 direct_message_event = event_dict_type(
diff --git a/zerver/tests/test_email_mirror.py b/zerver/tests/test_email_mirror.py
index b05fdf8610..abe7a61e97 100644
--- a/zerver/tests/test_email_mirror.py
+++ b/zerver/tests/test_email_mirror.py
@@ -1077,7 +1077,7 @@ class TestMissedMessageEmailMessages(ZulipTestCase):
         incoming_valid_message["To"] = mm_address
         incoming_valid_message["Reply-to"] = self.example_email("cordelia")
 
-        with self.assert_database_query_count(23):
+        with self.assert_database_query_count(22):
             process_message(incoming_valid_message)
 
         # Confirm Iago received the message.
diff --git a/zerver/tests/test_muted_users.py b/zerver/tests/test_muted_users.py
index 114dca0fb6..717bafc5f7 100644
--- a/zerver/tests/test_muted_users.py
+++ b/zerver/tests/test_muted_users.py
@@ -366,6 +366,11 @@ class MutedUsersTests(ZulipTestCase):
             m.assert_not_called()
 
     def test_muting_and_unmuting_restricted_users(self) -> None:
+        # NOTE: It is a largely unintended side effect of how we use
+        # the access_user_by_id API that limited guests can't mute
+        # inaccessible users. These tests verify the expected
+        # behavior.
+
         self.set_up_db_for_testing_user_access()
         polonius = self.example_user("polonius")
         cordelia = self.example_user("cordelia")
diff --git a/zerver/tests/test_users.py b/zerver/tests/test_users.py
index ba14d0e29a..84330c75ac 100644
--- a/zerver/tests/test_users.py
+++ b/zerver/tests/test_users.py
@@ -826,7 +826,7 @@ class QueryCountTest(ZulipTestCase):
 
         prereg_user = PreregistrationUser.objects.get(email="fred@zulip.com")
 
-        with self.assert_database_query_count(95):
+        with self.assert_database_query_count(94):
             with self.assert_memcached_count(23):
                 with self.capture_send_event_calls(expected_num_events=11) as events:
                     fred = do_create_user(
diff --git a/zerver/views/muted_users.py b/zerver/views/muted_users.py
index c0847f9956..42e7b97ffd 100644
--- a/zerver/views/muted_users.py
+++ b/zerver/views/muted_users.py
@@ -15,6 +15,14 @@ def mute_user(request: HttpRequest, user_profile: UserProfile, muted_user_id: in
     if user_profile.id == muted_user_id:
         raise JsonableError(_("Cannot mute self"))
 
+    # Arguably, access_used_by_id is not quite the right check; in the
+    # corner case of a limited guest trying to mute a deactivated user
+    # who they no longer have access to because the user was
+    # deactivated... it might from a policy perspective be OK to allow
+    # that operation even though this API will reject it.
+    #
+    # But it's quite possibly something nobody will try to do, so we
+    # just reuse the existing shared code path.
     muted_user = access_user_by_id(
         user_profile, muted_user_id, allow_bots=True, allow_deactivated=True, for_admin=False
     )
```

sahil839 commented:
Apart from the above mentioned follow-ups, a couple of more things are -
- Updating the push notifications and message events data to not include original details of inaccessible user. But this will only be possible for stream messages and when the sender is unsubscribed to the stream.

I have done some work on adding client capability. Even if we decide not to do that, we might still want to add code for web-app to not show the inaccessible users in sidebar and a couple of more minor changes like showing only limited info for inaccessible users in user popover.

timabbott commented:
Oh yeah, we definitely need that sort of web app changes before we can really turn this on.

timabbott commented:
I guess the early-exit in `zerver/actions/streams.py` failed coverage; just added a `nocoverage` comment so it can merge; probably worth fixing that in a follow-up PR but not a priority at all; I'm not worried about that `return` failing unexpectedly.
